File: 1250-longest-common-subsequence/longest-common-subsequence.js
/**
 * @param {string} text1
 * @param {string} text2
 * @return {number}
 */
var longestCommonSubsequence = function(text1, text2) {
    const memo = new Map()

    function lcs(i, j) {
        if (i === text1.length || j === text2.length) return 0

        const key = i + ',' + j

        if (memo.has(key)) return memo.get(key)

        let result

        if (text1[i] === text2[j]) {
            result = 1 + lcs(i + 1, j + 1)
        } else {
            result = Math.max(
                lcs(i + 1, j),
                lcs(i, j + 1)
            )
        }

        memo.set(key, result)
        return result
    }

    return lcs(0, 0)
};

File: 1250-longest-common-subsequence/longest-common-subsequence.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        prev_row = [0] * (len(text2) + 1)
        for i in range(len(text1) - 1, -1, -1):
            curr_row = [0] * (len(text2) + 1)
            for j in range(len(text2) - 1, -1, -1):
                if text1[i] == text2[j]:
                    curr_row[j] = 1 + prev_row[j + 1]

                else:
                    curr_row[j] = max(prev_row[j], curr_row[j+1])

            prev_row = curr_row
        return prev_row[0]


'''

i a e c
j
a 2 1 1 0
c 1 1 1 0
b 0 0 0 0
  0 0 0
'''